rj_gameplay/rj_gameplay/tactic/wall_tactic.py

```diff
@@ -0,0 +1,199 @@
+
+def find_wall_pts(num_robots: int, ball_pt: np.ndarray, goal_pt: np.ndarray, world_state: rc.WorldState):
+    """Calculates num_robots points to form a wall between the ball and goal.
+    :return list of wall_pts (numpy)
+    """
+    # TODO: param server this const
+    WALL_SPACING = BallConstants.RADIUS * 1.9
+
+    # dist is slightly greater than penalty box bounds
+    box_w = world_state.field.penalty_long_dist_m
+    box_h = world_state.field.penalty_short_dist_m
+    line_w = world_state.field.line_width_m
+    DIST_FROM_DEF = RobotConstants.RADIUS + line_w + np.hypot(box_w/2, box_h)
+
+    # check if vision is up and running
+    # (if it is these points should not be equal)
+    if (ball_pt == goal_pt).all():
+        return None
+
+    # get direction vec
+    dir_vec = (ball_pt - goal_pt) / np.linalg.norm(ball_pt - goal_pt)
+    wall_vec = np.array([dir_vec[1], -dir_vec[0]])
+
+    mid_pt = goal_pt + (dir_vec * DIST_FROM_DEF)
+    wall_pts = [mid_pt]
+
+    # set wall points in middle out pattern, given wall dir vector and WALL_SPACING constant
+    wall_pts = [mid_pt]
+    for i in range(num_robots-1):
+        mult = i//2 + 1
+        delta = (mult * (2 * RobotConstants.RADIUS + WALL_SPACING)) * wall_vec 
+        if i % 2: delta = -delta
+        wall_pts.append(mid_pt + delta)
+
+    return wall_pts
+
+def match_robots_to_wall(num_robots: int, ball_pt: np.ndarray, goal_pt: np.ndarray, world_state: rc.WorldState):
+    """Matches all wall_points to a robot.
+    :return dictionary of robot_id to List(rc.Robot, pt)
+    """
+    # assignments[robot_id] = [rc.Robot, pt]
+    assignments = {}
+
+    wall_pts = find_wall_pts(num_robots, ball_pt, goal_pt, world_state)
+    if not wall_pts: 
+        # sometimes tactic is called before ball_pos is seen
+        return None
+
+    # assign every wall pt a robot to move there
+    for wall_pt in wall_pts:
+        robot = robot_to_wall_pt(wall_pt, world_state, assignments)
+        assignments[robot.id] = (robot, wall_pt)
+
+    return assignments
+
+def robot_to_wall_pt(wall_pt: np.ndarray, world_state: rc.WorldState, assignments):
+    # TODO: dict type for assignments arg?
+    """Chooses which robot to move to a specific wall pt.
+    :return closest Robot object by dist
+    """
+    # find closest robot by dist, return it
+    min_robot = None
+    min_dist = float('inf')
+    for robot in world_state.our_robots:
+        # prevent duplicate assignments
+        if robot.id in assignments: 
+            continue
+
+        robot_pt = robot.pose[0:2]
+        dist = np.linalg.norm(robot_pt - wall_pt)
+        if dist < min_dist:
+            min_dist = dist
+            min_robot = robot
+
+    return min_robot
+
+class WallTactic(tactic.ITactic):
+
+    def __init__(self, 
+            num_robots: int, 
+            ball_pt: np.ndarray = None, 
+            goal_pt: np.ndarray = None,
+        ):
+
+        self.num_robots = num_robots
+        self.ball_pt = ball_pt 
```

Comment:
I don't, originally I had it named something like "mark_pt" because the old wall tactic took any two arbitrary points to form a wall between. Then I text replaced it all to ball_pt bc I didn't like that. I'll fix it.